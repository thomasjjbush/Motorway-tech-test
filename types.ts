import { FormEvent, ReactNode } from 'react';
import HtmlWebPackPlugin from 'html-webpack-plugin';
import { Json } from 'enzyme-to-json';

export interface AspectRatioProps {
    children: ReactNode;
    ratio: AspectRatios;
}

export enum AspectRatios {
    '1:1' = 100,
    '16:9' = 56.25,
    '4:3' = 75,
}

export interface ButtonProps {
    children: string;
    disabled?: boolean;
    onClick?: () => void;
    type?: 'button' | 'submit' | 'reset';
}

export interface FormField {
    label: string;
    max?: number;
    min?: number;
    placeholder?: string;
    type?: string;
    valid?: boolean;
    validation?: RegExp;
    value?: number | string;
}

export interface FormProps {
    ariaLabel: string;
    fields: {
        [field: string]: FormField;
    };
    onSubmit: (e: FormEvent) => void;
}

export interface Image {
    id: string;
    created_at: string;
    updated_at: string;
    color: string;
    description: string;
    alt_description: string;
    categories: string[];
    likes: number;
    user: {
        id: string;
        updated_at: string;
        username: string;
        name: string;
        first_name: string;
        last_name: string;
        bio: string;
        location: string;
        profile_image: string;
        total_collections: number;
        total_likes: number;
        total_photos: number;
    };
    url: string;
}

export interface ImageProps {
    alt: string;
    round?: boolean;
    src: string;
    width?: string;
}

export interface ImagesProps {
    columns: number;
}

export interface SerializerMap extends Json {
    node: {
        type: {
            componentStyle?: {
                baseStyle?: {
                    rules: (string | ((props: StyledProps) => string))[];
                };
                rules: (string | ((props: StyledProps) => string))[];
            };
        };
    };
}

export type StyledProps<P = unknown> = { theme: Theme } & P;

export interface Theme {
    colors: {
        blue: string;
        red: string;
        yellow: string;
    };
}

export interface WebpackConfig {
    devServer: { historyApiFallback: boolean; open: boolean };
    entry: string[];
    mode: string;
    module: {
        rules: WebpackRule[];
    };
    output: { path?: string; publicPath: string };
    plugins: any[];
    resolve: { extensions: string[] };
}

interface WebpackRule {
    exclude?: (RegExp | string)[];
    test: RegExp;
    use: {
        loader: string;
        options?: { plugins: string[] };
    };
}
