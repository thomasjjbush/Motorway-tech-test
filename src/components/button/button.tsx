import React, { FC, ReactElement } from 'react';
import styled from 'styled-components';
import { ButtonProps as Props, StyledProps } from '../../../types';

const StyledButton = styled.button<StyledProps>`
    appearance: none;
    background-color: ${({ theme }) => theme.colors.blue};
    border: 0;
    border-radius: 5px;
    color: white;
    cursor: pointer;
    font-size: 16px;
    padding: 10px 20px;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.5;
    }
`;

export const Button: FC<Props> = ({ children, disabled = false, onClick, type = 'button' }: Props): ReactElement => (
    <StyledButton disabled={disabled} onClick={onClick} type={type}>
        {children}
    </StyledButton>
);
