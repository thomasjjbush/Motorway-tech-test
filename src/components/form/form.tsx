import React, { ChangeEvent, FC, ReactElement, useState } from 'react';
import styled from 'styled-components';
import { FormProps as Props, StyledProps } from '../../../types';
import { Button } from '../button/button';

const Label = styled.label`
    display: block;
    margin-bottom: 5px;
`;

const Input = styled.input<StyledProps<{ valid: boolean }>>`
    border-radius: 5px;
    height: 30px;
    margin-bottom: 20px;
    padding: 0 10px;

    ${({ theme, valid }): string => `border: solid 2px ${valid ? 'black' : theme.colors.red};`}
`;

const StyledForm = styled.form`
    border: solid 2px black;
    border-radius: 5px;
    margin: 20px auto 0;
    max-width: 1000px;
    padding: 30px;
`;

export const Form: FC<Props> = ({ ariaLabel, fields: initialFields, onSubmit }: Props): ReactElement => {
    const [fields, setFields] = useState(initialFields);

    const onChange = (key: string, e: ChangeEvent<HTMLInputElement>) => {
        setFields({
            ...fields,
            [key]: {
                ...fields[key],
                valid: fields[key].validation?.test(e.currentTarget.value) ?? true,
                value: e.currentTarget.value,
            },
        });
    };

    return (
        <StyledForm aria-label={ariaLabel} data-test-id="form" onSubmit={onSubmit}>
            {Object.entries(fields).map(([key, { label, max, min, type, valid = false, validation, value = '' }]) => (
                <div key={key}>
                    <Label htmlFor={key}>{label}</Label>
                    <Input
                        id={key}
                        max={max}
                        min={min}
                        name={label}
                        onChange={(e) => onChange(key, e)}
                        type={type}
                        valid={Boolean(!validation && value) || valid}
                        value={value}
                    />
                </div>
            ))}
            <Button
                disabled={Object.values(fields).some(({ valid, validation }) => validation && !valid)}
                type="submit"
            >
                Submit
            </Button>
        </StyledForm>
    );
};
