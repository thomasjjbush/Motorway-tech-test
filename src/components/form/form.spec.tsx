import React from 'react';
import { shallow } from 'enzyme';
import { FormProps as Props } from '../../../types';
import { Form } from './form';
import { Button } from '../button/button';

describe('Form', () => {
    let props: Props;

    beforeEach(() => {
        props = {
            ariaLabel: 'aria label',
            fields: {
                mock: { label: 'mock', validation: /mockValue/ },
                anotherMock: { label: 'Another mock', type: 'number', min: 0, max: 30, value: 5 },
            },
            onSubmit: jest.fn(),
        };
    });

    it('should match snapshot', () => {
        expect(shallow(<Form {...props} />)).toMatchSnapshot();
    });

    it('should update value and not update valid onChange if value does not pass validation', () => {
        const wrapper = shallow(<Form {...props} />);
        expect(wrapper.find({ id: 'mock' }).prop('valid')).toBe(false);

        wrapper.find({ id: 'mock' }).prop('onChange')({ currentTarget: { value: 'mockValu3' } });
        expect(wrapper.find({ id: 'mock' }).prop('value')).toBe('mockValu3');
        expect(wrapper.find({ id: 'mock' }).prop('valid')).toBe(false);
    });

    it('should update value and apply valid prop to input onChange if value passes validation', () => {
        const wrapper = shallow(<Form {...props} />);
        expect(wrapper.find({ id: 'mock' }).prop('valid')).toBe(false);

        wrapper.find({ id: 'mock' }).prop('onChange')({ currentTarget: { value: 'mockValue' } });
        expect(wrapper.find({ id: 'mock' }).prop('value')).toBe('mockValue');
        expect(wrapper.find({ id: 'mock' }).prop('valid')).toBe(true);
    });

    it('should not apply disabled to submit button if all fields are valid and invoke onSubmit cb on click', () => {
        const wrapper = shallow(<Form {...props} fields={{ mock: { label: 'mock', value: 'mockValue' } }} />);
        expect(wrapper.find(Button).prop('disabled')).toBe(false);

        wrapper.find({ 'data-test-id': 'form' }).prop('onSubmit')();
        expect(props.onSubmit).toHaveBeenCalledTimes(1);
    });
});
