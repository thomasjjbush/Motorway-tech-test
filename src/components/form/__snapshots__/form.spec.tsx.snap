// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match snapshot 1`] = `
<styled.form
  aria-label="aria label"
  data-test-id="form"
  onSubmit={[MockFunction]}
  styles="
    border: solid 2px black;
    border-radius: 5px;
    margin: 20px auto 0;
    max-width: 1000px;
    padding: 30px;
"
>
  <div
    key="mock"
  >
    <styled.label
      htmlFor="mock"
      styles="
    display: block;
    margin-bottom: 5px;
"
    >
      mock
    </styled.label>
    <styled.input
      id="mock"
      name="mock"
      onChange={[Function]}
      styles="
    border-radius: 5px;
    height: 30px;
    margin-bottom: 20px;
    padding: 0 10px;

    border: solid 2px theme.colors.red;
"
      valid={false}
      value=""
    />
  </div>
  <div
    key="anotherMock"
  >
    <styled.label
      htmlFor="anotherMock"
      styles="
    display: block;
    margin-bottom: 5px;
"
    >
      Another mock
    </styled.label>
    <styled.input
      id="anotherMock"
      max={30}
      min={0}
      name="Another mock"
      onChange={[Function]}
      styles="
    border-radius: 5px;
    height: 30px;
    margin-bottom: 20px;
    padding: 0 10px;

    border: solid 2px black;
"
      type="number"
      valid={true}
      value={5}
    />
  </div>
  <Button
    disabled={true}
    type="submit"
  >
    Submit
  </Button>
</styled.form>
`;
